### PR TITLE
FEATURE - Rate limit by key

### PR DESCRIPTION
**Pull request by xiejiangzhi**
### Summary

Rate limit by a custom key. Migrate a app to Kong is easier, we can use rate limit with app authorization

### Full changelog

* Add `key` to limit_by and add `key_name`, `key_in_body` to rate limit. Allow limit by this key, we can find the key from header, query and body

### CODE REVIEW

File: .gitignore
Comment (thibaultcha) [inline .gitignore]: Please remove such unrelated changes from your commit - thank you!

File: kong/plugins/rate-limiting/handler.lua
Comment (thibaultcha) [inline kong/plugins/rate-limiting/handler.lua]: [kong.request.get_header()](https://docs.konghq.com/latest/pdk/kong.request/#kong_request_get_header) and [kong.request.get_query_arg()](https://docs.konghq.com/latest/pdk/kong.request/#kong-request-get_query_arg) would be more appropriate here.

File: kong/plugins/rate-limiting/handler.lua
Comment (thibaultcha) [inline kong/plugins/rate-limiting/handler.lua]: Please write the above as:
```lua
local body, err = kong.request.get_body()
if err then
  ngx_log(...)
  return
end

identifier = ...
```
 
As per our style guide and rules around early-exit branches. Also, error logs should _not_ be capitalized. Thanks!

File: kong/plugins/rate-limiting/schema.lua
Comment (thibaultcha) [inline kong/plugins/rate-limiting/schema.lua]: This schema should also ensure that `key_in_body` cannot be set if `key` is not the current policy?

File: kong/plugins/rate-limiting/schema.lua
Comment (thibaultcha) [inline kong/plugins/rate-limiting/schema.lua]: These schema updates will need to be ported to the new format _after_ https://github.com/Kong/kong/pull/3778 is merged. It converts our plugins schemas to the new schema library we have developed for Services and Routes, and is becoming the new schema library for Kong 1.0.

**Review by thibaultcha — CHANGES_REQUESTED** [review]
@xiejiangzhi Thank you for the patch! Left a few comments. Also, we will have to wait for https://github.com/Kong/kong/pull/3778 to be merged first, as described in my comments :)

File: kong/plugins/rate-limiting/handler.lua
Comment (thibaultcha) [reply kong/plugins/rate-limiting/handler.lua]: Right. I should have expanded the diff! Nonetheless, we should avoid the empty line between the call and the error handling :)

File: kong/plugins/rate-limiting/schema.lua
Comment (xiejiangzhi) [reply kong/plugins/rate-limiting/schema.lua]: The `key_in_body` is invalid if `key` is not the current policy.  It will be used when policy is `key` only. 
Do you have any advice?

Comment (thibaultcha) [reply kong/plugins/rate-limiting/schema.lua]: Yes, I am advising that if `key_in_body` is configured while `limit_by` is not `key`, then `self_check` returns an error; because as you said, `key_in_body` is invalid if `key` is not the current limiter.

In the new schema library, this would translate with what we call "entity checkers", that can ensure that a field if set if and only if _another_ field is set a particular way. See an example [here](https://github.com/Kong/kong/blob/next/kong/plugins/rate-limiting/schema.lua#L26-L27). Of course, this PR still has to remain in a pending state until the new schemas are merged into master, at which point, the above behavior can be implemented in this PR.

**Comment by thibaultcha** [comment]
Updated this PR with the "WIP" label to reflect the need to wait for #3778 to be merged before moving forward here. Thank you for your understanding!

File: kong/plugins/rate-limiting/schema.lua
Comment (xiejiangzhi) [reply kong/plugins/rate-limiting/schema.lua]: Thanks, I'll update it after the new schema is merged.

**Comment by james-callahan** [comment]
FYI #3778 has now been merged; are you able to rebase+update this PR?

**Comment by xiejiangzhi** [comment]
Hi @james-callahan #3778 just merged into `next` branch, so, should I rebase based on `next` and change this pr merge into `next`?

**Comment by xiejiangzhi** [comment]
Done

File: kong/plugins/rate-limiting/schema.lua
Comment (james-callahan) [inline kong/plugins/rate-limiting/schema.lua]: There's a couple of misc whitespace changes in this PR.

File: spec/03-plugins/24-rate-limiting/01-schema_spec.lua
Comment (james-callahan) [inline spec/03-plugins/24-rate-limiting/01-schema_spec.lua]: This function doesn't exist (see failing travis run):
```
Checking spec/03-plugins/24-rate-limiting/01-schema_spec.lua 4 warnings
    spec/03-plugins/24-rate-limiting/01-schema_spec.lua:20:24: accessing undefined variable validate_entity
    spec/03-plugins/24-rate-limiting/01-schema_spec.lua:20:48: accessing undefined variable plugin_schema
    spec/03-plugins/24-rate-limiting/01-schema_spec.lua:47:26: accessing undefined variable validate_entity
    spec/03-plugins/24-rate-limiting/01-schema_spec.lua:47:50: accessing undefined variable plugin_schema
```

File: kong/plugins/rate-limiting/schema.lua
Comment (james-callahan) [inline kong/plugins/rate-limiting/schema.lua]: nit: misc whitespace

File: spec/03-plugins/24-rate-limiting/01-schema_spec.lua
Comment (james-callahan) [inline spec/03-plugins/24-rate-limiting/01-schema_spec.lua]: nit: misc whitespace

**Review by james-callahan — CHANGES_REQUESTED** [review]
Just had a quick look over and commented on a couple of things that jumped out at me. I haven't thought about this PR in a wider scope yet.

File: kong/plugins/rate-limiting/handler.lua
Comment (james-callahan) [inline kong/plugins/rate-limiting/handler.lua]: Could you use `kong.log.err` here?
https://docs.konghq.com/0.14.x/pdk/kong.log/#kong-log-level

**Comment by xiejiangzhi** [comment]
Also has some problem in the schema, I think it isn't my side.
the access testing is passed.

```
 if_field = "limit_by", if_match = { eq = "key" },
 then_field = "key_name", then_match = { required = true },
```
and 

```
 if_field = "config.limit_by", if_match = { eq = "key" },
 then_field = "config.key_name", then_match = { required = true },
```

has a different result

I need some help.

**Comment by xiejiangzhi** [comment]
It seems that the testing has some errors. I saw `local ok, _, err = v(config, schema_def)` and `local ok, err = v(config, schema_def)` in your code. according to that helper code, it should be 2 arguments, right?

**Comment by codifierr** [comment]
Can this be used for rate-limiting oauth2 refresh token api?

**Pull request closed without merging**
